File: algo_expert/by_topic/stacks/easy/IsValid.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

void main(String[] args) {
    System.out.println(isValid("(){}}{")); // false
    System.out.println(isValid("((")); // false
    System.out.println(isValid("([{}])")); // true
    System.out.println(isValid("()[]")); // true
    System.out.println(isValid("([]])")); // false
}

private boolean isValid(String s) {
    if (s.length() < 2) return false;
    Stack<Character> stack = new Stack<>();
    int idx = 1;
    stack.push(s.charAt(0));
    Map<Character, Character> brackets = new HashMap<>();
    brackets.put(')', '(');
    brackets.put(']', '[');
    brackets.put('}', '{');

    while (idx < s.length()) {
        Character currentChar = s.charAt(idx);
        if (!brackets.containsKey(currentChar)) {
            stack.push(currentChar);
        } else {
            if (stack.isEmpty()) return false;
            Character bracket = stack.pop();
            if (!brackets.get(currentChar).equals(bracket)) return false;
        }
        idx++;
    }
    return stack.isEmpty();
}

File: algo_expert/by_topic/stacks/easy/isValid.js
/**
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * An input string is valid if:
 *
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Every close bracket has a corresponding open bracket of the same type.
 * Example 1:
 * Input: s = "()"
 * Output: true
 *
 * Example 2:
 * Input: s = "()[]{}"
 * Output: true
 *
 * Example 3:
 * Input: s = "(]"
 * Output: false
 *
 * Constraints:
 * 1 <= s.length <= 104
 * s consists of parentheses only '()[]{}'.
 */

function isValid(s) {
    let stack = [s[0]];
    let idx = 1;
    let brackets = {
        ")": "(",
        "]": "[",
        "}": "{"
    }

    while(stack.length > 0) {
        if (!(s[idx] in brackets)) {
            stack.push(s[idx]);
            idx++;
            continue;
        }
        let char = stack.pop();
        if (brackets[char] !== s[idx]) return false;
        idx++;
    }

    return true;
};